### PR TITLE
Reject direct injection for tuners with shared adapter state

### PR DESCRIPTION
**Pull request by CoralGarden52**
Related to #3631

## Summary

- Add a BaseTuner.uses_shared_state capability marker for tuners that keep adapter state shared across multiple target layers.
- Reject direct injection for TinyLoRA, UniLoRA, VeRA, PVeRA, VBLoRA, and FRoD before the model is modified.
- Keep get_peft_model support unchanged and document the limitation in the low-level API guide.

inject_adapter_in_model returns the mutated base model, while these tuners store shared adapter state on the tuner object. The direct API therefore cannot reliably support state-dict round-tripping. This change fails early with a clear error and points users to get_peft_model.

## Tests

- 12 passed: shared-state low-level API tests
- 385 passed, 1 xfailed: core tuner and low-level API tests
- 39 passed: shared-tuner low-memory loading tests
- make quality passed
- Full make test was started; the local run was stopped after an unrelated numerical failure in tests/test_custom_models.py::TestPeftCustomModel::test_merge_layers[Conv1d LOHA 3-Conv1dBigger-LoHaConfig-config_kwargs91] (torch.allclose during LoHa Conv1D merge).

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@CoralGarden52 Thanks for the PR. As it's in draft state, I assume it's not ready for review yet. If it is, mark it as ready and ping me please.

**Comment by CoralGarden52** [comment]
Hi @BenjaminBossan, thanks for checking. I’ve marked the PR as ready for review now.

The implementation follows your suggested class-level capability marker approach. It adds `BaseTuner.uses_shared_state`, rejects direct injection for the affected shared-state tuners before modifying the model, and keeps `get_peft_model` supported.

The targeted tests and quality checks pass. The unrelated local LoHa Conv1D numerical failure is documented in the PR description.

Could you please take another look when convenient?

File: tests/test_low_level_api.py [hunk 1]
```diff
@@ -62,6 +68,35 @@ def forward(self, input_ids):
...
+    pytest.param(UniLoraConfig, {"r": 2, "theta_d_length": 16}, id="UniLoRA"),
+    pytest.param(VeraConfig, {"r": 2}, id="VeRA"),
+    pytest.param(PveraConfig, {"r": 2}, id="PVeRA"),
+    pytest.param(VBLoRAConfig, {"r": 2, "num_vectors": 8, "vector_length": 2}, id="VBLoRA"),
+    pytest.param(FrodConfig, {"progressbar": False}, id="FRoD"),
+]
+
+
+@pytest.mark.parametrize("config_cls, config_kwargs", SHARED_STATE_TUNER_CONFIGS)
+def test_inject_adapter_in_model_rejects_shared_state_tuners(config_cls, config_kwargs):
```
Comment (BenjaminBossan) [inline tests/test_low_level_api.py]: Could you please move this tests into `TestLowLevelFunctional`, add them after the existing tests.

File: tests/test_low_level_api.py [hunk 2]
```diff
@@ -62,6 +68,35 @@ def forward(self, input_ids):
...
+    original_module_names = [name for name, _ in model.named_modules()]
+
+    with pytest.raises(ValueError, match="shared state.*get_peft_model"):
+        inject_adapter_in_model(config, model)
+
+    assert [name for name, _ in model.named_modules()] == original_module_names
+
+
+@pytest.mark.parametrize("config_cls, config_kwargs", SHARED_STATE_TUNER_CONFIGS)
+def test_get_peft_model_supports_shared_state_tuners(config_cls, config_kwargs):
```
Comment (BenjaminBossan) [inline tests/test_low_level_api.py]: No need for this test, if this would fail, a bunch of existing tests would be red in the CI.

Note that if you remove this test, there is no further need for a separate `SHARED_STATE_TUNER_CONFIGS` constant, so you can roll the list directly into the `parametrize` of `test_inject_adapter_in_model_rejects_shared_state_tuners`

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for providing this fix and corresponding tests. I have a few comments, please check.

What's missing for me is another test though: Imagine a new PEFT method with shared state is added that forgets to set `uses_shared_state = True`. As is, we wouldn't notice this, same as we didn't notice for the existing methods. Therefore, we need to add one more test that checks the specific roundtrip failure you describe in #3631.

As to where to put this: Normally, `test_low_level_api.py` would be the best place, as this is the API we want to test. However, in that file, we don't have a full matrix of all PEFT methods. Adding it there would be too much. Thus, I think the next best place to put this test is in `test_custom_models.py`. Use `set_init_weights_false` to ensure that the PEFT adapter is not an identity transform at the start and has no effect. LMK if you have questions.

File: tests/test_low_level_api.py (hunk 1 above)
Comment (CoralGarden52) [reply tests/test_low_level_api.py]: I moved the shared-state rejection test into `TestLowLevelFunctional` and placed it immediately after the existing `test_inject_adapter_in_model` test.

File: tests/test_low_level_api.py (hunk 2 above)
Comment (CoralGarden52) [reply tests/test_low_level_api.py]: I removed the redundant `test_get_peft_model_supports_shared_state_tuners` test. The shared-state configuration list is no longer defined as a separate constant and is now inlined directly into the parametrization of `test_inject_adapter_in_model_rejects_shared_state_tuners`.

The existing test matrix already covers the `get_peft_model` path.

**Comment by CoralGarden52** [comment]
Hi @BenjaminBossan,

I added the requested full-matrix regression test in `tests/test_custom_models.py`.

The test uses `TEST_CASES` and `set_init_weights_false`, injects the adapter directly into the source model, obtains its state dict with `get_peft_model_state_dict`, then injects a fresh model with a different initialization seed and restores it with `set_peft_model_state_dict`. It checks for unexpected keys and verifies that the outputs are identical after the round-trip.

For tuners already marked with `uses_shared_state = True`, the expected `ValueError` is accepted. If a future shared-state tuner forgets to set this marker, direct injection will proceed and the round-trip assertions will catch the missing shared state described in #3631.

The full matrix passes with 331 tests passed.

File: tests/test_custom_models.py [hunk 3]
```diff
@@ -2448,6 +2449,41 @@ def test_save_pretrained(self, test_name, model_id, config_cls, config_kwargs):
...
+            base_model_name_or_path=model_id,
+            **config_kwargs,
+        )
+
+        model = self.transformers_class.from_pretrained(model_id).to(self.torch_device)
+        torch.manual_seed(0)
+        try:
+            model = inject_adapter_in_model(config, model)
+        except ValueError as error:
+            if "shared state" not in str(error) or "get_peft_model" not in str(error):
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py:2468]: Let's add a comment here to explain what is checked and why we early return.

File: tests/testing_common.py [hunk 4]
```diff
@@ -296,6 +297,12 @@ def _test_load_model_low_cpu_mem_usage(self, model_id, config_cls, config_kwargs
             # also test injecting directly
             del model
             model = self.transformers_class.from_pretrained(model_id).to(self.torch_device)
+            tuner_cls = PEFT_TYPE_TO_TUNER_MAPPING[config.peft_type]
+            if tuner_cls.uses_shared_state:
+                with pytest.raises(ValueError, match="shared state.*get_peft_model"):
```
Comment (BenjaminBossan) [inline tests/testing_common.py:304]: Same, add a comment to explain.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for moving the test and adding the new one. My only issue left is about explaining the test a bit better. Otherwise the PR looks good.

File: tests/test_custom_models.py (hunk 3 above)
Comment (CoralGarden52) [reply tests/test_custom_models.py:2468]: I added a comment explaining that shared-state tuners are expected to reject direct injection, so the test returns because there is no round-trip to perform in that case.

The comment also explains why the error message is checked explicitly: unrelated ValueErrors must still fail the test, while a future tuner that forgets to set `uses_shared_state = True` will continue to the round-trip and expose the missing shared state.

File: tests/testing_common.py (hunk 4 above)
Comment (CoralGarden52) [reply tests/testing_common.py:304]: I added a comment explaining that direct injection is intentionally unsupported for shared-state tuners, so the expected ValueError completes this branch. The meta-device assertion below applies only to tuners that support direct injection.

**Comment by CoralGarden52** [comment]
Hi @BenjaminBossan ,

I added comments explaining the test behavior, including why shared-state tuners are expected to raise and why the test returns early.

When you have a chance, could you please take another look? If you have any further suggestions, I’m happy to make the necessary changes.

File: tests/test_custom_models.py (hunk 3 above)
Comment (BenjaminBossan) [reply tests/test_custom_models.py:2468]: Thanks for adding the comment. You can remove the "if a future tuner" part, this should be evident.

Comment (CoralGarden52) [reply tests/test_custom_models.py:2468]: Hi @BenjaminBossan ,

I removed the “if a future tuner forgets the marker” part and kept the comment focused on the expected rejection for shared-state tuners and the explicit error check.

The change is included in commit `a6059955`. When you have a chance, please take another look. I’m happy to make any further changes you suggest.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for adding this check that direct injection does not silently go through for PEFT methods that require shared weights with the on the parent class. LGTM.
